common/src/main/java/io/crate/interval/IntervalParser.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.interval;
+
+import org.joda.time.Period;
+
+import java.math.BigDecimal;
+
+public class IntervalParser {
+
+    public enum Precision {
+        YEAR,
+        MONTH,
+        DAY,
+        HOUR,
+        MINUTE,
+        SECOND
+    }
+
+    private IntervalParser() { }
+
+    public static Period apply(String value) {
+        return apply(value, null, null);
+    }
+
+    public static Period apply(String value, Precision start, Precision end) {
+        Period result;
+        try {
+            result = NumericalIntervalParser.apply(value, start, end);
+        } catch (IllegalArgumentException e1) {
+            try {
+                result = IsoIntervalParser.apply(value, start, end);
+            } catch (IllegalArgumentException e2) {
+                try {
+                    result = SQLStandardIntervalParser.apply(value, start, end);
+                } catch (IllegalArgumentException e3) {
+                    result = PGIntervalParser.apply(value, start, end);
+                }
+            }
+        }
+        return result.normalizedStandard();
+    }
+
+    static Period roundToPrecision(Period period, Precision start, Precision end) {
+        if (start == null && end == null) {
+            return period;
+        }
+        if (start == Precision.YEAR && end == null) {
+            return Period.years(period.getYears());
+        }
+        if (start == Precision.YEAR && end == Precision.MONTH) {
+            return Period.years(period.getYears()).withMonths(period.getMonths());
+        }
+        if (start == Precision.MONTH && end == null) {
+            return Period.years(period.getYears()).withMonths(period.getMonths());
+        }
+        if (start == Precision.DAY && end == null) {
+            return new Period(period).withHours(0).withMinutes(0).withSeconds(0).withMillis(0);
+        }
+        if (start == Precision.DAY && end == Precision.HOUR) {
```

Comment:
How about having a single `if (start == Precision.DAY)` and then within that branch handle the different `end` cases to avoid repeated `start == DAY` checks.

Reply:
Yep it won't be shorter, but it safes some CPU cycles and personally I think it is also easier to read, as you don't have to mentally repeatedly parse 2 expressions.